### PR TITLE
Fix a bug around trailing dots in field names.

### LINKED ISSUE #21862: array_index_out_of_bounds_exception for field name which is "."
**Elasticsearch version**: 5.0.1

**Plugins installed**: [] none

**JVM version**:1.8.0_111

**OS version**: Ubuntu 16.04

**Description of the problem including expected versus actual behavior**:

**Steps to reproduce**:
Without any custom templates or mapping, try this:

```
POST /test/test/1
{
    "." : {
        "foo" : {
            "bar" : 123
        }
    }
}
```

and you will receive this 400 error:
```
{
  "error" : {
    "root_cause" : [
      {
        "type" : "mapper_parsing_exception",
        "reason" : "failed to parse"
      }
    ],
    "type" : "mapper_parsing_exception",
    "reason" : "failed to parse",
    "caused_by" : {
      "type" : "array_index_out_of_bounds_exception",
      "reason" : "-1"
    }
  },
  "status" : 400
}
```

It seems that a single `.` is not a valid field name (I cannot find documentation detailing this), and even if this is the case, the error is not very helpful.

### PR DESCRIPTION
**Pull request by jtibshirani**
Previously, we didn't correctly handle trailing dots in field names:
* If the field name was only composed of dots like `.`, then we threw an
  `ArrayIndexOutOfBoundsException`.
* For other field names like `field.` we accepted the field and silently
  stripped away the dot.

We always threw the right exception for field names with leading dots like
`.field`.

This PR disallows trailing dots in field names for indices created in 8.0. For
7.x indices, it makes sure to maintain the old behavior for backwards
compatibility. It also clarifies the error messages that are returned.

### CODE REVIEW

**Comment by jtibshirani** [comment]
@elasticmachine run elasticsearch-ci/packaging-sample-matrix

File: server/src/main/java/org/elasticsearch/index/mapper/ObjectMapper.java [hunk 1]
```diff
@@ -277,7 +280,21 @@ protected static void parseProperties(ObjectMapper.Builder objBuilder, Map<Strin
                     if (typeParser == null) {
                         throw new MapperParsingException("No handler for type [" + type + "] declared on field [" + fieldName + "]");
                     }
-                    String[] fieldNameParts = fieldName.split("\\.");
+
+                    // We pass explicitly pass the 'limit' parameter to String#split to make sure that the last component is always
+                    // returned, even when it's an empty string. This allows us to handle trailing dots in field names correctly.
+                    String[] fieldNameParts = fieldName.split("\\.", -1);
```
Comment (cbuescher) [inline server/src/main/java/org/elasticsearch/index/mapper/ObjectMapper.java]: Would it be possible to detect empty string parts here already and issue a more comprehensive error message (maybe similar to the IAE that the FieldMapper throws now, but containing the full `fieldName` to make finding the offending field easier e.g. in a large mapping?

**Review by cbuescher — COMMENTED** [review]
@jtharpla I took a look and left an idead around whether it would be possible to enhance the error message a bit. Let me know what you think about the idea, the rest looks good to me already.

**Comment by jtibshirani** [comment]
@cbuescher thanks for the review! I realized that this doesn't directly address #21862 as I claimed, since that relates to dynamic mapping updates. I'll get up to speed on that issue before returning to this PR, in case it makes sense to tackle static and dynamic mapping updates at the same time.

**Comment by cbuescher** [comment]
@jtibshirani wondering if we should close this PR since it seems outdated, wdyt?

**Comment by jtibshirani** [comment]
@cbuescher thanks for the ping, I'll close this and open a new one if I end up picking it up again.

**Pull request closed without merging**